DS13/code/modules/overmap/components.dm

```diff
@@ -0,0 +1,643 @@
+	icon_state = "injector"
+	req_access = list(ACCESS_ENGINE_EQUIP)
+	var/obj/item/tank/internals/plasma/loaded_tank = null //Nicked from rad collector code, see there for documentation (collector.dm)
+	max_integrity = 350
+	integrity_failure = 50
+	var/powerproduction_drain = 0.015 //Relatively high drain on the plasma tanks.
+	var/drainratio = 1
+	var/supplying = FALSE //Are we supplying plasma to another object?
+	var/obj/structure/overmap_component/plasma_relay/supply_to
+	var/obj/structure/overmap_component/integrity_field_generator/generator
+	var/locked = FALSE
+
+/obj/structure/overmap_component/plasma_injector/attack_hand(mob/user)
+	var/mob/living/carbon/human/L = user
+	var/obj/item/card/id/ID = L.wear_id
+	if(ID && istype(ID))
+		if(!check_access(ID))
+			to_chat(user, "<span class='boldnotice'>Unable to comply</span> - <span class='warning'>you are not authorized to execute this command.</span>")
+			return
+	else
+		to_chat(user, "You need to be wearing an ID to use this machine.")
+		return
+	eject()
+
+/obj/structure/overmap_component/plasma_injector/examine(mob/user)
+	. = ..()
+	if(!supply_to)
+		return
+	to_chat(user, "it's injecting plasma towards [supply_to] in [get_area(supply_to)].")
+
+/obj/structure/overmap_component/plasma_injector/attackby(obj/item/W, mob/user, params)
+	if(istype(W, /obj/item/tank/internals/plasma))
+		if(!anchored)
+			to_chat(user, "<span class='warning'>[src] needs to be secured to the floor first!</span>")
+			return TRUE
+		if(loaded_tank)
+			to_chat(user, "<span class='warning'>There's already a plasma tank loaded!</span>")
+			return TRUE
+		if(!user.transferItemToLoc(W, src))
+			return
+		loaded_tank = W
+		START_PROCESSING(SSobj,src)
+	else if(W.GetID())
+		if(allowed(user))
+			locked = !locked
+			to_chat(user, "<span class='notice'>You [locked ? "lock" : "unlock"] the controls.</span>")
+		else
+			to_chat(user, "<span class='danger'>Access denied.</span>")
+			return TRUE
+	else
+		return ..()
+
+/obj/structure/overmap_component/plasma_injector/Destroy()
+	STOP_PROCESSING(SSobj,src)
+	. = ..()
+
+/obj/structure/overmap_component/plasma_injector/process()
+	if(!loaded_tank || !linked || !linked.powered_components.len)
+		icon_state = "injector"
+		return
+	if(!supply_to || QDELETED(supply_to))
+		supplying = FALSE
+		find_supply_to()
+		return
+	if(!loaded_tank.air_contents.gases[/datum/gas/plasma])
+		playsound(src, 'DS13/sound/effects/computer/alert2.ogg', 100, 1)
+		playsound(src, 'DS13/sound/effects/computer/power_down.ogg', 100, 1)
+		eject()
+		icon_state = "injector"
+		return
+	icon_state = "injector-on"
+	var/gasdrained = min(powerproduction_drain*drainratio,loaded_tank.air_contents.gases[/datum/gas/plasma][MOLES])
+	loaded_tank.air_contents.gases[/datum/gas/plasma][MOLES] -= gasdrained
+	loaded_tank.air_contents.garbage_collect()
+	if(supply_to)
+		supply_to.plasma_volume += powerproduction_drain
+	if(generator)
+		generator.plasma_volume += powerproduction_drain
```

Comment:
So it generates twice as much energy if you have both a generator and something to supply it to?

Reply:
skkkeeeeeeeeeeeeeeeeeeet